### ISSUE TITLE
Is there any reason you don't use torch.nn.utils.spectral_norm()?

### ISSUE THREAD

**Issue by Zhangyanbo** [jarrelscy/iResnet#5]
Pytorch has it built-in spectral normalization, i.e., the torch.nn.utils.spectral_norm(). Is there any reason you don't use that?

**Comment by jarrelscy** [comment]
The inbuilt Pytorch spectral norm uses Miyato's version of Spectral
Normalisation (for GANs) (https://arxiv.org/pdf/1802.05957.pdf). The
iresnet paper specifies that they use Gouk's version of spectral norm (
https://arxiv.org/pdf/1804.04368.pdf) which is implemented slightly
differently.

There are several differences but the largest and probably most relevant to
us is that for CNNs Miyato regularises the flattened kernel of the CNN
instead of computing the "true" spectral norm of the entire function (which
naively would require calculating a huge, mostly sparse matrix). Gouk's
method provides an efficient way of estimating the spectral norm through
the power method (https://en.wikipedia.org/wiki/Power_iteration).

**Comment by Zhangyanbo** [comment]
Thank you! This is really helpful!
